src/data_race.rs

```diff
@@ -961,18 +970,30 @@ trait EvalContextPrivExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
     /// FIXME: is this valid, or should get_raw_mut be used for
     /// atomic-stores/atomic-rmw?
     fn validate_atomic_op<A: Debug + Copy>(
-        &self,
+        &mut self,
         place: MPlaceTy<'tcx, Tag>,
         atomic: A,
         description: &str,
+        write: bool,
         mut op: impl FnMut(
             &mut MemoryCellClocks,
             &mut ThreadClockSet,
             VectorIdx,
             A,
         ) -> Result<(), DataRace>,
     ) -> InterpResult<'tcx> {
-        let this = self.eval_context_ref();
+        let this = self.eval_context_mut();
+
+        // Update yield metadata for yield based forward progress and live-lock detection.
+        let place_ptr = place.ptr.assert_ptr();
+        let size = place.layout.size;
+        if write {
+            this.thread_yield_atomic_wake(place_ptr.alloc_id, place_ptr.offset,size);
```

Comment:
Instead of passing `alloc_id` and `offset` separately, what about passing the entire `Pointer`?